ns/implementing authentication.

nicolesigei opened the pull request:
Adds feature #17 

**Features  proposed in this pull request**
Implementing Authentication by connecting the app to Microsoft Graph using OpenID Connect for sign in, Microsoft Authentication Library (MSAL v4) for .NET to obtain an access token, and the Microsoft Graph Client Library for .NET (SDK) to interact with Microsoft Graph.

nicolesigei committed:
Implementing Authentication

Connecting to Microsoft Graph using delegated permissions and Azure AD endpoint.

nicolesigei committed:
Integration of the app to SharePoint.

Connecting the app to SharePoint to get the Site using SiteID and getting Lists using ListID.

irvinesunday commented on AssetTracking/AssetTracking/Controllers/AdminController.cs:
```diff
@@ -0,0 +1,75 @@
+
+        [Authorize]
+        // Loads the user's profile.
+        public async Task<IActionResult> Index(string email)
```
Consider using [Role-based authorization](https://docs.microsoft.com/en-us/aspnet/core/security/authorization/roles?view=aspnetcore-2.2) for Administrators:

nicolesigei replied:
This shall be implemented once users are added, Thank You

irvinesunday commented on AssetTracking/AssetTracking/Controllers/HomeController.cs:
```diff
@@ -1,24 +1,38 @@
+        public GraphServiceClient GraphClient { get; private set; }
+
+        [AllowAnonymous]
+        public IActionResult Index(string email)
```
Is the `email` parameter really required?

irvinesunday commented on AssetTracking/AssetTracking/Extensions/AzureAdAuthenticationBuilderExtensions.cs:
```diff
@@ -0,0 +1,104 @@
+
+                        var currentTenantId = context.Principal.FindFirst(Startup.TenantIdType).Value;
+
+                        if (currentTenantId == "1b0b9aff-6c77-416c-8248-b3345736cfc9")
```
Is this code block necessary? What does it do?

irvinesunday commented on AssetTracking/AssetTracking/Helpers/GraphAuthProvider.cs:
```diff
@@ -0,0 +1,76 @@
+
+        public async Task<AuthenticationResult> GetUserAccessTokenByAuthorizationCode(string authorizationCode)
+        {
+
```
Remove this extra whitespace

irvinesunday commented on AssetTracking/AssetTracking/Helpers/GraphSDKHelper.cs:
```diff
@@ -0,0 +1,43 @@
+                    // Append the access token to the request
+                    requestMessage.Headers.Authorization = new AuthenticationHeaderValue("Bearer", accessToken);
+                   // This header identifies the sample in the Microsoft Graph service. If extracting this code for your project please remove.
+                    requestMessage.Headers.Add("SampleID", "aspnetcore-connect-sample");
```
Let's change these string values to reflect our own sample app.

irvinesunday commented on AssetTracking/AssetTracking/Helpers/GraphService.cs:
```diff
@@ -0,0 +1,189 @@
+        // Loads the user's profile in formatted JSON.
+        public static async Task<string> GetUserJson(GraphServiceClient graphClient, string email, HttpContext httpContext)
+        {
+            if (email == null) return JsonConvert.SerializeObject(new { Message = "Email address cannot be null." }, Formatting.Indented);
```
For ease of code readability and/or maintenance, have the `return` code in its own line or enclosed in double braces.

irvinesunday commented on AssetTracking/AssetTracking/Helpers/GraphService.cs:
```diff
@@ -0,0 +1,189 @@
+                {
+                    case "ResourceNotFound":
+                        // If picture not found, return the default image.
+                        return "data:image/svg+xml;base64,PD94bWwgdmVyc2lvbj0iMS4wIiBlbmNvZGluZz0iVVRGLTgiPz4NCjwhRE9DVFlQRSBzdmcgIFBVQkxJQyAnLS8vVzNDLy9EVEQgU1ZHIDEuMS8vRU4nICAnaHR0cDovL3d3dy53My5vcmcvR3JhcGhpY3MvU1ZHLzEuMS9EVEQvc3ZnMTEuZHRkJz4NCjxzdmcgd2lkdGg9IjQwMXB4IiBoZWlnaHQ9IjQwMXB4IiBlbmFibGUtYmFja2dyb3VuZD0ibmV3IDMxMi44MDkgMCA0MDEgNDAxIiB2ZXJzaW9uPSIxLjEiIHZpZXdCb3g9IjMxMi44MDkgMCA0MDEgN ...
```
Consider having the default photo stored in-memory in a folder and retrieved dynamically rather than having it string hard-coded in case one might want to change the default, and also to reduce the potential of a corrupted value.

irvinesunday commented on AssetTracking/AssetTracking/Helpers/GraphService.cs:
```diff
@@ -0,0 +1,189 @@
+                    if (e.Error.Code == "GetUserPhoto") // User is using MSA, we need to use beta endpoint
+                    {
+                        // Set Microsoft Graph endpoint to beta, to be able to get profile picture for MSAs 
+                        graphClient.BaseUrl = "https://graph.microsoft.com/beta";
```
Consider using the `GraphResourceId` [property ](https://github.com/microsoftgraph/aspnet-assettracking-sample/blob/ea4e13c937fa32f3eff082dbc2ce53b1cf0d92d3/AssetTracking/AssetTracking/Extensions/AzureAdOptions.cs#L13)here: 
`graphClient.BaseUrl = configuration.GraphResourceId + beta;`

You can also have the `beta` and `v1.0` values as constants so that they are defined once and re-used throughout this class.

irvinesunday commented on AssetTracking/AssetTracking/Helpers/GraphService.cs:
```diff
@@ -0,0 +1,189 @@
+                        pictureStream = await graphClient.Users[email].Photo.Content.Request().GetAsync();
+
+                        // Reset Microsoft Graph endpoint to v1.0
+                        graphClient.BaseUrl = "https://graph.microsoft.com/v1.0";
```
Same as above:
`graphClient.BaseUrl = configuration.GraphResourceId + version_1_0;`

Where `version_1_0` is a constant for "v1.0"

irvinesunday commented on AssetTracking/AssetTracking/Helpers/GraphService.cs:
```diff
@@ -0,0 +1,189 @@
+                    if (e.Error.Code == "GetUserPhoto") // User is using MSA, we need to use beta endpoint
+                    {
+                        // Set Microsoft Graph endpoint to beta, to be able to get profile picture for MSAs 
+                        graphClient.BaseUrl = "https://graph.microsoft.com/beta";
```
Do the same as above for this and the same instances below this line.

irvinesunday commented on AssetTracking/AssetTracking/Models/Sites.cs:
```diff
@@ -0,0 +1,166 @@
+{
+    public static class Sites
+    {
+        /// <summary>
```
Complete the parameter and summary descriptions in the code documentation blocks for this and the subsequent methods in this class.

irvinesunday commented on AssetTracking/AssetTracking/Models/Sites.cs:
```diff
@@ -0,0 +1,166 @@
+                              .Request()
+                              .AddAsync(listItem);
+            }
+            catch (Exception ex)
```
Provide the implementation for this `try catch` handler and the method below

irvinesunday commented on AssetTracking/AssetTracking/Models/Sites.cs:
```diff
@@ -0,0 +1,166 @@
+        /// <returns></returns>
+        public static async Task<ListItem> DeleteListItem(GraphServiceClient graphServiceClient, ListItem listItem, string groupId, string siteId, string listId)
+        {
+            throw new NotImplementedException();
```
Provide the implementation for this method

nicolesigei replied:
This is still a working progress

irvinesunday commented on AssetTracking/AssetTracking/Models/Sites.cs:
```diff
@@ -0,0 +1,166 @@
+            }
+            catch (ServiceException e)
+            {
+                return null;
```
We could return a more meaningful error here.

irvinesunday commented on AssetTracking/AssetTracking/Models/Sites.cs:
```diff
@@ -0,0 +1,166 @@
+        /// <returns></returns>
+        public static async Task<Site> GetSites(GraphServiceClient graphServiceClient, string groupsId, string siteId, string listId)
+        {
+            var sites = await graphServiceClient
```
Wrap this in a `try catch` block which return meaningful error messages. Do the same for the below resource methods which already don't have this implemented.

irvinesunday commented on AssetTracking/AssetTracking/appsettings.json:
```diff
@@ -1,8 +1,34 @@
+
+    "ClientId": "c2d47c42-f92d-49d1-9390-7fe44779f665",
+
+    "ClientSecret": "Ye[1Rpy.9Jv**u8rfUIprVLSxxKQH6Ct", // This sample uses a password (secret) to authenticate. Production apps should use a certificate.
```
Use ASP.NET Core client [secret manager](https://docs.microsoft.com/en-us/aspnet/core/security/app-secrets?view=aspnetcore-2.2&tabs=windows#secret-manager) to manage app. confidential info. (`ClientSecret` and `ClientId`)

nicolesigei replied:
The link doesn't work

nicolesigei replied:
Okay will read more into it.

titusgicheru commented on AssetTracking/AssetTracking/Startup.cs:
```diff
@@ -1,25 +1,36 @@
 using Microsoft.AspNetCore.Mvc;
 using Microsoft.Extensions.Configuration;
 using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Identity.Client;
```
I don't think the assembly references below are being used within the class, please remove them

using Microsoft.Identity.Client;
using Microsoft.IdentityModel.Protocols.OpenIdConnect;
using Microsoft.Owin.Logging;
using Microsoft.Owin.Security.Notifications;
using Microsoft.Owin.Security.OpenIdConnect;
using System.Configuration;
using System.Threading.Tasks;

titusgicheru commented on AssetTracking/AssetTracking/Startup.cs:
```diff
@@ -31,13 +42,35 @@ public void ConfigureServices(IServiceCollection services)
-        }
+            services.AddSession();
 
+            // Add application services.
```
Remove commented section line 58, 59

titusgicheru reviewed (changes requested):
- Please cleanup all unnecessary white spaces within the project classes and ensure proper indentation.
- Update the README.md file with the intro content from your design doc.

nicolesigei committed:
Clean up of all unnecessary whitespaces

Removed all unnecessary white spaces and classes.

irvinesunday commented on AssetTracking/AssetTracking/Extensions/AzureAdOptions.cs:
```diff
@@ -0,0 +1,17 @@
+﻿namespace AssetTracking.Extensions
+{
+    public class azureOptions
```
Let's capitalize the first letter of this class name so that we maintain naming standards for Classes.

nicolesigei committed:
Capitalization of class name

changed the class name from azureOptions to AzureOptions

nicolesigei committed:
Removing app confidential information.

Removing the user's confidential information that is the client secret and the client id

nicolesigei committed:
added launchsettings.json file

nicolesigei committed:
Removed duplicate Error methods on the Home Controller

nicolesigei committed:
Removed duplicate Error methods in the Home Controller

nicolesigei committed:
Merge remote-tracking branch 'origin/ns/Implementing-authentication-and-integration-to-sharepoint' into ns/Implementing-authentication-and-integration-to-sharepoint

nicolesigei committed:
Changed the Redirect URL in the Account Controller from Home Controller to Admin Controller

nicolesigei committed:
Changed the redirectUrl from Home Controller to Admin Controller

nicolesigei committed:
Renamed file.

The previous file name had a typo.

nicolesigei committed:
Renamed file.

The previous file name had a typo.

nicolesigei committed:
Changed the Redirect URI and used secret manager 

The secret manager was used to manage app confidential information like Client ID and the Client Secret.

nicolesigei committed:
Changed the Redirect URI and used secret manager 

The secret manager was used to manage app confidential information like Client ID and the Client Secret.

nicolesigei committed:
Merge remote-tracking branch 'origin/ns/Implementing-authentication-and-integration-to-sharepoint' into Implementing-authentication

nicolesigei committed:
Removing because of pending implementation.

This will be added to the Integration to SharePoint branch

nicolesigei committed:
Removing because of pending implementation

This will be added to the Integration to SharePoint branch
